### PR TITLE
Turns stickers into a component and an element

### PR DESCRIPTION
**Pull request by mc-oofert**
## About The Pull Request
Adds /datum/element/sticker, and /datum/component/attached_sticker
Sticker items now mostly operate off /datum/element/sticker
The sticker element, hooks the whole "attach to stuff" and adds the attached_sticker component to its target
The attached_sticker component, adds the overlay, hooks the clean and on-fire signals.

## Why It's Good For The Game

Allows to check if a sticker is present on an object (which I will use later)
Code is probably cleaner???

## Changelog
:cl:
refactor: Stickers use a component and an element now to do their sticking
/:cl:

### CODE REVIEW

File: code/datums/components/sticker.dm [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	sticker_overlay = mutable_appearance(stick.icon, stick.icon_state , layer = atom_parent.layer + 1, appearance_flags = RESET_COLOR | PIXEL_SCALE)
+	sticker_overlay.pixel_x = px
+	sticker_overlay.pixel_y = py
+	atom_parent.add_overlay(sticker_overlay)
+	if(isliving(parent) && user)
+		var/mob/living/victim = parent
+		if(victim.client)
+			user.log_message("stuck [sticker] to [key_name(victim)]", LOG_ATTACK)
+			victim.log_message("had [sticker] stuck to them by [key_name(user)]", LOG_ATTACK)
+	if(isturf(parent) && user)
```
Comment (Jacquerel) [inline code/datums/components/sticker.dm:71]: Might be worth doing an `if/else` for the type, as you can't be both living and a turf.

File: code/datums/components/sticker.dm [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+		//register signals on the users turf instead because we can assume they are on flooring sticking it to a wall so it should burn (otherwise it would fruitlessly check wall temperature)
+		signal_turf = (user && isclosedturf(parent)) ? get_turf(user) : parent
```
Comment (Jacquerel) [inline code/datums/components/sticker.dm:73]: You also checked for `user` above, so I'd either remove it from that case or from this one. 
You probably want to register the signal regardless of if there was a user, so keeping it on this line makes more sense.

File: code/datums/components/sticker.dm [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+		sticker = null
+	qdel(src)
+
+/datum/component/attached_sticker/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_LIVING_IGNITED, PROC_REF(on_ignite))
+	RegisterSignal(parent, COMSIG_COMPONENT_CLEAN_ACT, PROC_REF(peel))
+	RegisterSignal(parent, COMSIG_PARENT_QDELETING, PROC_REF(on_attached_qdel))
+
+/datum/component/attached_sticker/UnregisterFromParent()
+	SIGNAL_HANDLER
```
Comment (Jacquerel) [inline code/datums/components/sticker.dm:98]: As the linter says, this isn't a signal handler.

**Review by Jacquerel — CHANGES_REQUESTED** [review]

File: code/datums/components/sticker.dm (hunk 3 above)
Comment (mc-oofert) [reply code/datums/components/sticker.dm:98]: i gorgor

**Comment by optimumtact** [comment]
can anyone explain to me exactly what other item needs this behaviour that made it worth componentising?

**Comment by mc-oofert** [comment]
> can anyone explain to me exactly what other item needs this behaviour that made it worth componentising?

well no item yet cause the one that im trying to do kinda needs this behaviour as a component

also admins like spamming stuff with elements for some reason anyway

File: code/datums/components/attached_sticker.dm [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	if(sticker)
+		sticker.forceMove(isturf(parent) ? parent : as_atom.drop_location())
+		sticker.pixel_y = rand(-4,1)
+		sticker.pixel_x = rand(-3,3)
+		sticker = null
+	qdel(src)
+
+/datum/component/attached_sticker/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_LIVING_IGNITED, PROC_REF(on_ignite))
+	RegisterSignal(parent, COMSIG_COMPONENT_CLEAN_ACT, PROC_REF(peel))
```
[the remark this answers is not in the document]
Comment (mc-oofert) [reply code/datums/components/attached_sticker.dm:49]: the signal handlers already check if the item itself actually has the "flammable" resistance flag

[the remark this answers is not in the document]
Comment (mc-oofert) [reply code/datums/components/attached_sticker.dm:49]: well the fire reactions only

**Pull request merged**